newspaper/extractors.py

```diff
@@ -511,26 +511,35 @@ def get_canonical_link(self, article_url, doc):
         1. The rel=canonical tag
         2. The og:url tag
         """
+        links = self.parser.getElementsByTag(doc, tag='link', attr='rel',
+                                             value='canonical')
 
-        result = ''
-
-        links = self.parser.getElementsByTag(doc, tag='link', attr='rel', value='canonical')
         canonical = self.parser.getAttribute(links[0], 'href') if links else ''
-
         og_url = self.get_meta_content(doc, 'meta[property="og:url"]')
-
-        if canonical:
-            canonical = canonical.strip()
-            o = urllib.parse.urlparse(canonical)
-            if not o.hostname:
-                z = urllib.parse.urlparse(article_url)
-                domain = '%s://%s' % (z.scheme, z.hostname)
-                canonical = urllib.parse.urljoin(domain, canonical)
-            result = canonical
-        elif og_url:
-            result = og_url  # fallback to og:url tag, must be the full url
-
-        return result
+        meta_url = canonical or og_url or ''
+        if meta_url:
+            meta_url = meta_url.strip()
+            parsed_meta_url = urlparse(meta_url)
+            if not parsed_meta_url.hostname:
+                # MIGHT not have a hostname in meta_url
+                # parsed_url.path might be 'example.com/article.html' where
+                # clearly example.com is the hostname
+                parsed_article_url = urlparse(article_url)
+                strip_hostname_in_meta_path = re.\
+                            match(".*{}(?=/)/(.*)".
+                            format(parsed_article_url.hostname),
+                                  parsed_meta_url.path)
+                try:
+                    true_path = strip_hostname_in_meta_path.group(1)
+                except AttributeError:
+                    true_path = parsed_meta_url.path
+
+                # true_path may contain querystrings and fragments
+                meta_url = urlunparse((parsed_article_url.scheme,
+                                       parsed_article_url.hostname,true_path,
```

Comment:
I would have put true_path on its own line, but thanks for fixing my python style in this 👍